viewing_party/party.py

```diff
@@ -1,23 +1,121 @@
-# -----------------------------------------
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+    return user_data
 
-        
-# -----------------------------------------
-# ------------- WAVE 4 --------------------
-# -----------------------------------------
+# ---------------------------------------------------- WAVE 2 ----------------------------------------------------
 
-# -----------------------------------------
-# ------------- WAVE 5 --------------------
-# -----------------------------------------
+def get_watched_avg_rating(user_data):
+    if not user_data["watched"]:
+        return 0
+    
+    total_movie_ratings = 0
 
+    for movie_rating in user_data["watched"]:
+        total_movie_ratings += movie_rating["rating"]
+
+    average_movie_rating = total_movie_ratings / len(user_data["watched"])
+
+    return average_movie_rating
+
+def get_most_watched_genre(user_data):
+    frequency_of_genre_index = {}
+    max_frequency = 0
+    most_watched_genre = None
+
+    for single_movie in user_data["watched"]:
+        genre_type = single_movie.get("genre")
+        if genre_type:
+            frequency_of_genre_index[genre_type] = frequency_of_genre_index.get(genre_type,0) + 1
+            if frequency_of_genre_index[genre_type] > max_frequency:
+              max_frequency = frequency_of_genre_index[genre_type]
+              most_watched_genre = genre_type
+
+    return most_watched_genre
+
+# ---------------------------------------------------- WAVE 3 ----------------------------------------------------
+
+def get_unique_watched(user_data):
+    if not user_data["watched"]:
+        return []
+  
+    my_friends_unique = [movie["title"] for my_friends in user_data["friends"] 
+                         for movie in my_friends["watched"]]
+    
+    my_unique_movies = [movie for movie in user_data["watched"] 
+                        if movie["title"] not in my_friends_unique]   
+
+    return my_unique_movies
+
+def get_friends_unique_watched(user_data):
+    friends_unique_movies = []
+    my_movie_titles = [] 
+    
+    my_movie_titles = [movie["title"] for movie in user_data["watched"]]
+
+    for my_friend in user_data["friends"]:           
+        for movie in my_friend["watched"]:
+            if movie["title"] not in my_movie_titles and movie not in friends_unique_movies:
+                friends_unique_movies.append(movie) 
+
+    return list(friends_unique_movies)
+
+# ---------------------------------------------------- WAVE 4 ----------------------------------------------------
+
+def get_available_recs(user_data):
+    recommended_movies = []
+    friends_unique_movies = get_friends_unique_watched(user_data)
+
+    user_watched_titles = set()
+    
+    user_watched_titles = [movie['title'] for movie in user_data.get('watched', [])]

```

Comment:
Is there a reason why this was declared as a `set` and then converted into a list? I would actually say to keep it as a set since it could improve the efficiency of your code.

Reply:
fixed: will wait to push